lib/charms/karma_k8s/v0/karma_dashboard.py

```diff
@@ -128,10 +131,37 @@ class RelationManagerBase(Object):
         name (str): consumer's relation name
     """
 
-    def __init__(self, charm: CharmBase, relation_name):
+    def __init__(self, charm: CharmBase, relation_name, relation_role: RelationRole):
         super().__init__(charm, relation_name)
+        self.charm = charm
+        self._validate_relation(relation_name, relation_role)
         self.name = relation_name
 
+    def _validate_relation(self, relation_name: str, relation_role: RelationRole):
+        try:
+            if self.charm.meta.relations[relation_name].role != relation_role:
+                raise ValueError(
+                    "Relation '{}' in the charm's metadata.yaml must be '{}' "
+                    "to be managed by this library, but instead it is '{}'".format(
+                        relation_name,
+                        relation_role,
+                        self.charm.meta.relations[relation_name].role,
+                    )
+                )
+            if self.charm.meta.relations[relation_name].interface_name != INTERFACE_NAME:
+                raise ValueError(
+                    "Relation '{}' in the charm's metadata.yaml must use the '{}' interface "
+                    "to be managed by this library, but instead it is '{}'".format(
+                        relation_name,
+                        INTERFACE_NAME,
+                        self.charm.meta.relations[relation_name].interface_name,
+                    )
+                )
+        except KeyError:
+            raise ValueError(
+                "Relation '{}' is not in the charm's metadata.yaml".format(relation_name)
+            )
```

Comment:
There isn't a centralized way for charm libraries to claim that they own a specific name. If we adopt this approach as a pattern, we risk the situation where two libraries cannot be used in a charm simply because of a name collision, rather than any inherent incompatibilities between the libraries.

I think that I'd prefer to see libraries that take a relation name and interface name as an argument to a constructor, thus allowing charm authors to work around naming collisions. The constructors can use default names on those arguments for a "zero config" approach, but I think that it's important to avoid hard coding things like names in our libraries.

Reply:
We have similar things in [loki](https://github.com/canonical/loki-k8s-operator/blob/cb14570fd5de503aa9980423db2cd5f0499fa08b/lib/charms/loki_k8s/v0/loki_push_api.py#L1146), [grafana](https://github.com/canonical/grafana-k8s-operator/blob/5cc6999f64c2044f701a51b79e700dcd07c1bdde/lib/charms/grafana_k8s/v0/grafana_dashboard.py#L742), [prometheus](https://github.com/canonical/prometheus-k8s-operator/blob/98ec8278fda13bf23091ffab18d00b23ebceeada/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L980), [alertmanager](https://github.com/canonical/alertmanager-k8s-operator/blob/279f4bfc051c181f6ef91197320cedca1748c121/lib/charms/alertmanager_k8s/v0/alertmanager_dispatch.py#L92).

Can you please further explain the collision scenario?
The only hard coded name here is `INTERFACE_NAME = "karma_dashboard` and that cannot be changed if a charm wants to relate to this particular karma.

To my understanding the only scenario this limits is reusing this library as is for other purposes.

Side note: relations, and not charms, should be 1st-class citizens (assuming there can be only one; otherwise both relations and charms can be 1st class citizens :) ).